README.md
# EEPROM 28C Programmer eXPerience
This is a home project for a simple parallel EEPROM programmer based on Raspberry PI (model 3).
It currently supports 28C16 (2Kx8), 28C64 (8Kx8) and 28C256 (32Kx8) EEPROMs, but it will probably work with all 28C-family (having all the necessary address-bus pins).

![eeprom-28C-programmer_built](eeprom-28C-programmer_built.jpg)

## Specifications
The need of model 3 of Raspberry is only for the number of usable GPIOs
(but it may be possible to save a lot of pins using serial-to-parallel shift registers for the address bus).
The different voltages between Raspberry GPIOs (3.3V) and EEPROM pin (5V) is managed by a TXS0108 (8-bit bidirectional voltage-level translator).
The software currently supports single-byte reads/writes, write operation is followed by hard-coded delay to complete a write cycle.

### Hardware
Schematics and PCB layouts are designed with ExpressPCB free CAD software.

#### Schematic:
![board-schematic](hardware/eeprom-28C-programmer_sch.jpg)

### Software
The software is written in C and uses WiringPI for GPIO access, and the GNU Getopt function to automate the parsing of command-line options.

Below is the list of options with the relative allowed params, as shown in the help:

![eeprom-28C-programmer_usage](software/eeprom-28C-programmer_usage.jpg)

Warn! The software, in this release, does not check the value of the parameters, be careful.

## Changes
See file [CHANGES](CHANGES.md) for the project resources change logs


## Future Plans
See file [TODO](TODO.md) for the project future plans


## About
Author : Alessandro Fraschetti (mail: [gos95@gommagomma.net](mailto:gos95@gommagomma.net))


## License
This project is licensed under the [MIT license](LICENSE)


CHANGES.md
# CHANGES
This file contains the log of changes of **EEPROM 28C Programmer** eXPerience.


## "mercury" 1.2.0 2020-07-13 Feature release
- Added software "-a" option to set datafile as ascii vs binary


## "mercury" 1.1.0 2020-07-10 Feature release
- Added software "-p" option to print the bus pin number of the RasbperryPI GPIOs (wiringPI numbering)


## "mercury" 1.0.0 2020-07-08 Feature release
- Refactored software path-tree structure and code modules


## 0.1.0 2020-07-05 Draft release
- Created repository *xp-eeprom-28C-programmer* and
- Added README.md, CHANGES.md and TODO.md files
- Added software code for RaspberryPI-3
- Added schematic
- Tested on Atmel 28C16, 28C64 and 28C256


software/eeprog28.c
/*
 * eeprog28.c
 *
 *  ___ ___ ___ ___  ___  __  __                                                     
 * | __| __| _ \ _ \/ _ \|  \/  |___ _ __ _ _ ___  __ _ _ _ __ _ _ __  _ __  ___ _ _ 
 * | _|| _||  _/   / (_) | |\/| |___| '_ \ '_/ _ \/ _` | '_/ _` | '  \| '  \/ -_) '_|
 * |___|___|_| |_|_\\___/|_|  |_|   | .__/_| \___/\__, |_| \__,_|_|_|_|_|_|_\___|_|  
 *                                  |_|           |___/                              
 *
 * Author.....: Alessandro Fraschetti (mail: gos95@gommagomma.net)
 * Target.....: RaspberryPI
 * Version....: 1.0 2020/06/28
 * Description: EEPROM 28C-family programmer utility
 * URL........: https://github.com/gom9000/xp-eeprom-28C-programmer
 * License....: this program is under the terms of MIT License
 * Compile....: gcc eeprog28.c -o eeprog28 -lwiringPi
 * Notes......: ADDR BUS : A0-A14 = 8,9,7  0,2,3  12,13,14  30,21,22,23,24,25 (wiringPi pin numbers)
 *              DATA BUS : D0-D7  = 15,16,1  4,5  6,10,11
 *              CTRL BUS : CE=27, OE=28, WE=29
 *
 */


#include <ctype.h>
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <unistd.h>
#include <wiringPi.h>


#define BUS_SIZE(bus) (sizeof(bus)/sizeof(bus[0]))


void setup(void);
void setAddress(uint32_t address);
void setData(uint8_t data);
uint8_t getData(void);
void setDataBusDirection(uint8_t dir);
uint8_t readROM(uint32_t address);
void writeROM(uint32_t address, uint8_t data);
void waitForWriteCycle(uint32_t address, uint8_t data);
void eraseROM(uint32_t start, uint32_t length, uint8_t value);
int testROM(uint32_t start, uint32_t length, uint8_t value);
void dumpROM(uint32_t start, uint32_t length, FILE *outstream);
void setSDPMode(uint8_t mode);


uint8_t D[] = {15,16,1,4,5,6,10,11}; // D0-D7
uint8_t A[] = {8,9,7,0,2,3,12,13,14,30,21,22,23,24,25}; //A0-A14
uint8_t CE = 27;
uint8_t OE = 28;
uint8_t WE = 29;


void setup()
{
    for (uint8_t ii=0; ii<BUS_SIZE(A); ii++)
    {
        pinMode(A[ii], OUTPUT);
        digitalWrite(A[ii], LOW);
    }

    setDataBusDirection(INPUT);

    pinMode(CE, OUTPUT);
    pinMode(OE, OUTPUT);
    pinMode(WE, OUTPUT);

    digitalWrite(CE, HIGH);
    digitalWrite(OE, HIGH);
    digitalWrite(WE, HIGH);
}

void setAddress(uint32_t address)
{
    for (uint8_t ii=0; ii<BUS_SIZE(A); ii++)
    {
        digitalWrite(A[ii], address & 1);
        address >>= 1;
    }
}

void setData(uint8_t data)
{
    for (uint8_t ii=0; ii<BUS_SIZE(D); ii++)
    {
        digitalWrite(D[ii], data & 1);
        data >>= 1;
    }
}

uint8_t getData()
{
    uint8_t data = 0;

    for (uint8_t ii=BUS_SIZE(D); ii>0; ii--)
        data = (data << 1) | digitalRead(D[ii-1]);

    return data;
}

void setDataBusDirection(uint8_t dir)
{
    for (uint8_t ii=0; ii<BUS_SIZE(D); ii++)
        pinMode(D[ii], dir);
}

uint8_t readROM(uint32_t address)
{
    uint8_t data = 0;
    setDataBusDirection(INPUT);
    setAddress(address);
    digitalWrite(CE, LOW);
    digitalWrite(OE, LOW);
    delayMicroseconds(1);
    data = getData();
    digitalWrite(OE, HIGH);
    digitalWrite(CE, HIGH);

    return data;
}

void writeROM(uint32_t address, uint8_t data)
{
    setAddress(address);
    setDataBusDirection(OUTPUT);
    setData(data);
    digitalWrite(CE, LOW);
    digitalWrite(WE, LOW);
    delayMicroseconds(1);
    digitalWrite(WE, HIGH);
    digitalWrite(CE, HIGH);
}

void waitForWriteCycle(uint32_t address, uint8_t data)
{
//    while (readROM(address)&0x80 != data&0x80) printf("waiting... %d  %d\n", address, data);
    delay(10);
}

void eraseROM(uint32_t start, uint32_t length, uint8_t value)
{
    for (uint32_t address=start; address<start+length; address++)
    {
       writeROM(address, value);
       waitForWriteCycle(address, value);
    }
}

int testROM(uint32_t start, uint32_t length, uint8_t value)
{
    uint32_t count = 0;
    for (uint32_t address=start; address<start+length; address++)
       if (readROM(address) != value) count++;;

    return count;
}

void dumpROM(uint32_t start, uint32_t length, FILE *outstream)
{
    if (outstream == stdout && start%16)
    {
        printf("%.5X:  ", start-start%16);
        for (uint8_t ii=0; ii<start%16; ii++)
            printf("   ");
    }

    for (uint32_t address=start; address<start+length; address++)
    {
        if (outstream == stdout && !(address%16)) printf("%.5X:  ", address);
        fprintf(outstream, "%.2X", readROM(address));
        if (outstream == stdout) printf(" ");
        if (address%16 == 15) fprintf(outstream, "\r\n");
    }
    if (outstream == stdout && (start+length)%16) printf("\n");
}

void setSDPMode(uint8_t mode)
{
    if (mode)
    {
        writeROM(0x5555, 0xAA);
        writeROM(0x2AAA, 0x55);
        writeROM(0x5555, 0xA0);
    } else {
        writeROM(0x5555, 0xAA);
        writeROM(0x2AAA, 0x55);
        writeROM(0x5555, 0x80);
        writeROM(0x5555, 0xAA);
        writeROM(0x2AAA, 0x55);
        writeROM(0x5555, 0x20);
    } 
}


int main(int argc, char **argv)
{
    int address = 0;
    int c;
    unsigned char t_flag = 0;
    unsigned char d_flag = 0;
    unsigned char e_flag = 0;
    unsigned char w_flag = 0;
    unsigned char v_flag = 0;
    unsigned char l_flag = 0;
    unsigned char p_flag = 0;
    int length = 0;
    int start = 0;
    int mode = 0;
    int zerobyte = 0xFF;
    char *filename = NULL;
    FILE *fp = stdout;


    wiringPiSetup();
    setup();

    while ((c = getopt(argc, argv, "s:f:z:t:d:e:w::v::l:ph")) != -1)
    {
        switch (c)
        {
            case 's':
                sscanf(optarg, "%5X", &start);
                break;
            case 'f':
                filename = optarg;
                break;
            case 'z':
                sscanf(optarg, "%2X", &zerobyte);
                break;
            case 't':
                t_flag = 1;
                sscanf(optarg, "%5X", &length);
                break;
            case 'd':
                d_flag = 1;
                sscanf(optarg, "%5X", &length);
                break;
            case 'e':
                e_flag = 1;
                sscanf(optarg, "%5X", &length);
                break;
            case 'w':
                w_flag = 1;
                if (optarg) sscanf(optarg, "%5X", &length);
                break;
            case 'v':
                v_flag = 1;
                if (optarg) sscanf(optarg, "%5X", &length);
                break;
            case 'l':
                l_flag = 1;
                if (optarg) sscanf(optarg, "%d", &mode);
                break;
            case 'p':
                p_flag = 1;
                break;
            case 'h':
                printf("usage: %s option [params...]\n", argv[0]);
                printf("EEPROM 28C programmer utility\n");
                printf("  options:\n");
                printf("    -t: <LENGTH> [-s <START>] [-z <BLANK>]\n");
                printf("    \ttest if rom is filled with BLANK data values\n");
                printf("    -d: <LENGTH> [-s <START>] [-z <BLANK>] [-f <DATAFILE>]\n");
                printf("    \tdump LENGTH bytes of rom, starting from START address, into DATAFILE\n");
                printf("    -e: <LENGTH> [-s <START>]\n");
                printf("    \terase LENGTH bytes (filled with BLANK values) of rom, starting from START address\n");
                printf("    -v: [<LENGTH>] [-s <START>] [-z <BLANK>] [-f <DATAFILE>]\n");
                printf("    \tverify LENGTH bytes of rom, starting from START address, with the contents of DATAFILE\n");
                printf("    \tif LENGTH is not specified or LENGTH is greater than DATAFILE length, the latter is used.\n");
                printf("    -w: [<LENGTH>] [-s <START>] [-z <BLANK>] [-f <DATAFILE>]\n");
                printf("    \twrite LENGTH bytes of rom, starting from START address, with the contents of DATAFILE\n");
                printf("    \tif LENGTH is not specified, DATAFILE length is used. If LENGTH is greter than DATAFILE length, diff is filled with BLANK values\n");
                printf("    -l: <MODE>\n");
                printf("    \tenable (MODE=1) or disable (MODE=0) rom Software Data Protection\n");
                printf("    -h: show this help and exit\n");
                printf("  params:\n");
                printf("    -s: <START> set the start hex address, default value is 0x0\n");
                printf("    -f: <DATAFILE> set the datafile where to dump/read hex values, default value for dump option is stdout\n");
                printf("    -z: <BLANK> set the blank hex data value to fill/test the rom, default value is 0xFF\n");
                printf("    -p: show the value of input params\n");
                return 0;
            case '?':
                if (optopt == 'w' || optopt == 'v')
                    fprintf(stderr, "Option -%c requires valid filename as param.\n", optopt);
                else if (optopt == 't' || optopt == 'd' || optopt == 'e')
                    fprintf(stderr, "Option -%c requires ROM size as param.\n", optopt);
                else if (isprint (optopt))
                    fprintf(stderr, "Unknown option `-%c'.\n", optopt);
                else
                    fprintf(stderr, "Unknown option character `\\x%x'.\n", optopt);
                return 1;
            default:
                abort();
        }
    }


    if (p_flag)
    {
        printf("start-address=%.5X, data-length=%.5X, zero-byte=%.2X, data-file=%s SDP=%d\n", start, length, zerobyte, filename, mode);
    }
    if (t_flag)
    {
        uint32_t count;
        if (count = testROM(start, length, zerobyte))
            printf("ROM is not empty, found %d data bytes\n", count);
        else
            printf("ROM is empty\n");
        return 0; 
    }
    if (d_flag)
    {
        if (filename) fp = fopen(filename, "w");
        printf("Dumping of 0x%X locations of ROM from start 0x%X...\n", length, start);
        dumpROM(start, length, fp);
        if (filename) fclose(fp);
        printf("ROM dumped.\n");
        return 0;
    }
    if (e_flag)
    {
        printf("Erasing (0x%X) of 0x%X locations of ROM from start 0x%X...\n", zerobyte, length, start);
        eraseROM(start, length, zerobyte);
        printf("ROM erased.\n");
        return 0;
    }
    if (w_flag)
    {
        int data = 0;
        uint32_t address = 0;
        if (filename) fp = fopen(filename, "r");
        printf("Writing ROM with contents of file: %s from start 0x%X...\n", filename, start);
        while (fscanf(fp, "%2X", &data) == 1 && address<length)
        {
            writeROM(start+address, data);
            waitForWriteCycle(start+address, data);
            address++;
        }
        if (filename) fclose(fp);
        if (address<length)
            eraseROM(start+address, length-address, zerobyte);
        printf("0x%X locations of ROM wrote.\n", address);
        return 0;
    }
    if (v_flag)
    {
        int data = 0;
        int actual = 0;
        uint32_t address = 0;
        if (filename) fp = fopen(filename, "r");
        if (!length) length = -1;
        printf("Verifying ROM with contents of file: %s from start 0x%X...\n", filename, start);
        while (fscanf(fp, "%2X", &data) == 1 && address-length)
        {
            actual = readROM(start+address++);
            if (data != actual)
            {
                data = -1;
                break;
            }
        } 
        if (filename) fclose(fp);
        if (data == -1)
            printf("verify error, found 0x%X at location 0x%X.\n", actual, --address);
        else
            printf("verify done.\n");
        return 0;
    }
    if (l_flag)
    {
        printf("ROM Software Data Protection %s.\n", (mode? "Enabled" : "Disabled"));
        setSDPMode(mode);
        return 0;
    }
    return 0;
}


software/src/eeprom.c
/*
 * eeprom.c
 *
 *  ___ ___ ___ ___  ___  __  __                                                     
 * | __| __| _ \ _ \/ _ \|  \/  |___ _ __ _ _ ___  __ _ _ _ __ _ _ __  _ __  ___ _ _ 
 * | _|| _||  _/   / (_) | |\/| |___| '_ \ '_/ _ \/ _` | '_/ _` | '  \| '  \/ -_) '_|
 * |___|___|_| |_|_\\___/|_|  |_|   | .__/_| \___/\__, |_| \__,_|_|_|_|_|_|_\___|_|  
 *                                  |_|           |___/                              
 *
 * Author.....: Alessandro Fraschetti (mail: gos95@gommagomma.net)
 * Target.....: RaspberryPI
 * Version....: 1.0 2020/07/06
 * Description: EEPROM 28C-family programmer utility
 * URL........: https://github.com/gom9000/xp-eeprom-28C-programmer
 * License....: this program is under the terms of MIT License
 * Module.....: eeprom operations library
 * Notes......: ADDR BUS : A0-A14 = 8,9,7  0,2,3  12,13,14  30,21,22,23,24,25 (wiringPi pin numbers)
 *              DATA BUS : D0-D7  = 15,16,1  4,5  6,10,11
 *              CTRL BUS : CE=27, OE=28, WE=29
 *
 */


#include <stdio.h>
#include <wiringPi.h>
#include "eeprom.h"


uint8_t D[] = {15,16,1,4,5,6,10,11}; // D0-D7
uint8_t A[] = {8,9,7,0,2,3,12,13,14,30,21,22,23,24,25}; //A0-A14
uint8_t CE = 27;
uint8_t OE = 28;
uint8_t WE = 29;


void setup()
{
    for (unsigned char ii=0; ii<BUS_SIZE(A); ii++)
    {
        pinMode(A[ii], OUTPUT);
        digitalWrite(A[ii], LOW);
    }

    setDataBusDirection(INPUT);

    pinMode(CE, OUTPUT);
    pinMode(OE, OUTPUT);
    pinMode(WE, OUTPUT);

    digitalWrite(CE, HIGH);
    digitalWrite(OE, HIGH);
    digitalWrite(WE, HIGH);
}

void setAddress(address_t address)
{
    for (unsigned char ii=0; ii<BUS_SIZE(A); ii++)
    {
        digitalWrite(A[ii], address & 1);
        address >>= 1;
    }
}

void setData(data_t data)
{
    for (unsigned char ii=0; ii<BUS_SIZE(D); ii++)
    {
        digitalWrite(D[ii], data & 1);
        data >>= 1;
    }
}

data_t getData()
{
    data_t data = 0;

    for (unsigned char ii=BUS_SIZE(D); ii>0; ii--)
        data = (data << 1) | digitalRead(D[ii-1]);

    return data;
}

void setDataBusDirection(unsigned char dir)
{
    for (unsigned char ii=0; ii<BUS_SIZE(D); ii++)
        pinMode(D[ii], dir);
}

data_t readROM(address_t address)
{
    data_t data = 0;
    setDataBusDirection(INPUT);
    setAddress(address);
    digitalWrite(CE, LOW);
    digitalWrite(OE, LOW);
    delayMicroseconds(1);
    data = getData();
    digitalWrite(OE, HIGH);
    digitalWrite(CE, HIGH);

    return data;
}

void writeROM(address_t address, data_t data)
{
    setAddress(address);
    setDataBusDirection(OUTPUT);
    setData(data);
    digitalWrite(CE, LOW);
    digitalWrite(WE, LOW);
    delayMicroseconds(1);
    digitalWrite(WE, HIGH);
    digitalWrite(CE, HIGH);
}

void waitForWriteCycle(address_t address, data_t data)
{
//    while (readROM(address)&0x80 != data&0x80) printf("waiting... %d  %d\n", address, data);
    delay(10);
}

void eraseROM(address_t start, length_t length, data_t blank)
{
    for (address_t address=start; address<start+length; address++)
    {
       writeROM(address, blank);
       waitForWriteCycle(address, blank);
    }
}

length_t testROM(address_t start, length_t length, data_t data)
{
    length_t count = 0;
    for (address_t address=start; address<start+length; address++)
       if (readROM(address) != data) count++;;

    return count;
}

void dumpROM(address_t start, length_t length, FILE *outstream)
{
    if (outstream == stdout && start%16)
    {
        printf("%.5X:  ", start-start%16);
        for (unsigned char ii=0; ii<start%16; ii++)
            printf("   ");
    }

    for (address_t address=start; address<start+length; address++)
    {
        if (outstream == stdout && !(address%16)) printf("%.5X:  ", address);
        fprintf(outstream, "%.2X", readROM(address));
        if (outstream == stdout) printf(" ");
        if (address%16 == 15) fprintf(outstream, "\r\n");
    }
    if (outstream == stdout && (start+length)%16) printf("\n");
}

void setSDPMode(unsigned char mode)
{
    if (mode)
    {
        writeROM(0x5555, 0xAA);
        writeROM(0x2AAA, 0x55);
        writeROM(0x5555, 0xA0);
    } else {
        writeROM(0x5555, 0xAA);
        writeROM(0x2AAA, 0x55);
        writeROM(0x5555, 0x80);
        writeROM(0x5555, 0xAA);
        writeROM(0x2AAA, 0x55);
        writeROM(0x5555, 0x20);
    } 
}


software/src/eeprom.h
/*
 * eeprom.h
 *
 *  ___ ___ ___ ___  ___  __  __                                                     
 * | __| __| _ \ _ \/ _ \|  \/  |___ _ __ _ _ ___  __ _ _ _ __ _ _ __  _ __  ___ _ _ 
 * | _|| _||  _/   / (_) | |\/| |___| '_ \ '_/ _ \/ _` | '_/ _` | '  \| '  \/ -_) '_|
 * |___|___|_| |_|_\\___/|_|  |_|   | .__/_| \___/\__, |_| \__,_|_|_|_|_|_|_\___|_|  
 *                                  |_|           |___/                              
 *
 * Author.....: Alessandro Fraschetti (mail: gos95@gommagomma.net)
 * Target.....: RaspberryPI
 * Version....: 1.1 2020/06/06
 * Description: EEPROM 28C-family programmer utility
 * URL........: https://github.com/gom9000/xp-eeprom-28C-programmer
 * License....: this program is under the terms of MIT License
 * Module.....: header file of library module eeprom.c
 * Notes......:
 */


#ifndef __EEPROM_H__
#define __EEPROM_H__


#include <stdint.h>
typedef uint32_t address_t;
typedef uint32_t length_t;
typedef uint8_t data_t;


#define BUS_SIZE(bus) (sizeof(bus)/sizeof(bus[0]))


extern uint8_t D[8]; // D0-D7
extern uint8_t A[15]; // A0-A14
extern uint8_t CE;
extern uint8_t OE;
extern uint8_t WE;


void setup(void);
void setAddress(address_t);
void setData(data_t);
data_t getData(void);
void setDataBusDirection(unsigned char);
data_t readROM(address_t);
void writeROM(address_t, data_t);
void waitForWriteCycle(address_t, data_t);
void eraseROM(address_t, address_t, data_t);
length_t testROM(address_t, length_t, data_t);
void dumpROM(address_t, length_t, FILE*);
void setSDPMode(unsigned char);


#endif /* !defined(__EEPROM_H__) */


software/src/main.c
/*
 * eeprog28.c
 *
 *  ___ ___ ___ ___  ___  __  __                                                     
 * | __| __| _ \ _ \/ _ \|  \/  |___ _ __ _ _ ___  __ _ _ _ __ _ _ __  _ __  ___ _ _ 
 * | _|| _||  _/   / (_) | |\/| |___| '_ \ '_/ _ \/ _` | '_/ _` | '  \| '  \/ -_) '_|
 * |___|___|_| |_|_\\___/|_|  |_|   | .__/_| \___/\__, |_| \__,_|_|_|_|_|_|_\___|_|  
 *                                  |_|           |___/                              
 *
 * Author.....: Alessandro Fraschetti (mail: gos95@gommagomma.net)
 * Target.....: RaspberryPI
 * Version....: 1.0 2020/06/28
 * Description: EEPROM 28C-family programmer utility
 * URL........: https://github.com/gom9000/xp-eeprom-28C-programmer
 * License....: this program is under the terms of MIT License
 * Notes......: ADDR BUS : A0-A14 = 8,9,7  0,2,3  12,13,14  30,21,22,23,24,25 (wiringPi pin numbers)
 *              DATA BUS : D0-D7  = 15,16,1  4,5  6,10,11
 *              CTRL BUS : CE=27, OE=28, WE=29
 *
 */


#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <wiringPi.h>
#include "eeprom.h"


int main(int argc, char **argv)
{
    int c;
    unsigned char t_flag = 0;
    unsigned char d_flag = 0;
    unsigned char e_flag = 0;
    unsigned char w_flag = 0;
    unsigned char v_flag = 0;
    unsigned char l_flag = 0;
    unsigned char a_flag = 0;
    unsigned char p_flag = 0;
    unsigned int start = 0;
    unsigned int length = 0;
    unsigned char mode = 0;
    unsigned int zerobyte = 0xFF;
    char *filename = NULL;
    FILE *fp = stdout;


    wiringPiSetup();
    setup();

    while ((c = getopt(argc, argv, "s:f:z:t:d:e:w::v::l:aph")) != -1)
    {
        switch (c)
        {
            case 's':
                sscanf(optarg, "%5X", &start);
                break;
            case 'f':
                filename = optarg;
                break;
            case 'z':
                sscanf(optarg, "%2X", &zerobyte);
                break;
            case 't':
                t_flag = 1;
                sscanf(optarg, "%5X", &length);
                break;
            case 'd':
                d_flag = 1;
                sscanf(optarg, "%5X", &length);
                break;
            case 'e':
                e_flag = 1;
                sscanf(optarg, "%5X", &length);
                break;
            case 'w':
                w_flag = 1;
                if (optarg) sscanf(optarg, "%5X", &length);
                break;
            case 'v':
                v_flag = 1;
                if (optarg) sscanf(optarg, "%5X", &length);
                break;
            case 'l':
                l_flag = 1;
                if (optarg) sscanf(optarg, "%c", &mode);
                break;
            case 'a':
                a_flag = 1;
                break;
            case 'p':
                p_flag = 1;
                break;
            case 'h':
                printf("usage: %s option [params...]\n", argv[0]);
                printf("EEPROM 28C programmer utility\n");
                printf("  options:\n");
                printf("    -t: <LENGTH> [-s <START>] [-z <BLANK>]\n");
                printf("    \ttest if rom is filled with BLANK data values\n");
                printf("    -d: <LENGTH> [-s <START>] [-z <BLANK>] [-f <DATAFILE>]\n");
                printf("    \tdump LENGTH bytes of rom, starting from START address, into DATAFILE\n");
                printf("    -e: <LENGTH> [-s <START>]\n");
                printf("    \terase LENGTH bytes (filled with BLANK values) of rom, starting from START address\n");
                printf("    -v: [<LENGTH>] [-s <START>] [-z <BLANK>] [-f <DATAFILE>]\n");
                printf("    \tverify LENGTH bytes of rom, starting from START address, with the contents of DATAFILE\n");
                printf("    \tif LENGTH is not specified or LENGTH is greater than DATAFILE length, the latter is used.\n");
                printf("    -w: [<LENGTH>] [-s <START>] [-z <BLANK>] [-f <DATAFILE>]\n");
                printf("    \twrite LENGTH bytes of rom, starting from START address, with the contents of DATAFILE\n");
                printf("    \tif LENGTH is not specified, DATAFILE length is used. If LENGTH is greter than DATAFILE length, diff is filled with BLANK values\n");
                printf("    -l: <MODE>\n");
                printf("    \tenable (MODE=1) or disable (MODE=0) rom Software Data Protection\n");
                printf("    -h: show this help and exit\n");
                printf("  params:\n");
                printf("    -s: <START> set the start hex address, default value is 0x0\n");
                printf("    -f: <DATAFILE> set the datafile where to dump/read hex values, default value for dump option is stdout\n");
                printf("    -z: <BLANK> set the blank hex data value to fill/test the rom, default value is 0xFF\n");
                printf("    -a: set the mode of the datafile as ascii (default mode is binary)\n");
                printf("    -p: show the value of input params and bus pin numbers\n");
                return 0;
            case '?':
                if (optopt == 'w' || optopt == 'v')
                    fprintf(stderr, "Option -%c requires valid filename as param.\n", optopt);
                else if (optopt == 't' || optopt == 'd' || optopt == 'e')
                    fprintf(stderr, "Option -%c requires ROM size as param.\n", optopt);
                else if (isprint (optopt))
                    fprintf(stderr, "Unknown option `-%c'.\n", optopt);
                else
                    fprintf(stderr, "Unknown option character `\\x%x'.\n", optopt);
                return 1;
            default:
                abort();
        }
    }


    if (p_flag)
    {
        printf("start-address=%.5X, data-length=%.5X, zero-byte=%.2X, data-file=%s, data-file-mode=%s, SDP=%d\n", start, length, zerobyte, filename, (a_flag? "ascii":"binary"), mode);
        printf("address bus pins: ");
        for (unsigned char ii=0; ii<BUS_SIZE(A); ii++) printf("%d ", A[ii]);
        printf("\ndata bus pins: ");
        for (unsigned char ii=0; ii<BUS_SIZE(D); ii++) printf("%d ", D[ii]);
        printf("\nWE pin: %d, CE pin: %d, OE pin: %d\n", WE, CE, OE);
    }
    if (t_flag)
    {
        length_t count;
        if ((count = testROM(start, length, zerobyte)))
            printf("ROM is not empty, found %d data bytes\n", count);
        else
            printf("ROM is empty\n");
        return 0; 
    }
    if (d_flag)
    {
        if (filename) fp = fopen(filename, "w");
        printf("Dumping of 0x%X locations of ROM from start 0x%X...\n", length, start);
        dumpROM(start, length, fp);
        if (filename) fclose(fp);
        printf("ROM dumped.\n");
        return 0;
    }
    if (e_flag)
    {
        printf("Erasing (0x%X) of 0x%X locations of ROM from start 0x%X...\n", zerobyte, length, start);
        eraseROM(start, length, zerobyte);
        printf("ROM erased.\n");
        return 0;
    }
    if (w_flag && !a_flag)
    {
        data_t data = 0;
        address_t address = 0;
        if (filename) fp = fopen(filename, "rb");
        printf("Writing ROM with contents of binary file: %s from start 0x%X...\n", filename, start);
        while (!feof(fp) && address<length)
        {
            fread(&data, sizeof(data_t), 1, fp); 
            writeROM(start+address, data);
            waitForWriteCycle(start+address, data);
            address++;
        }
        if (filename) fclose(fp);
        if (address<length)
            eraseROM(start+address, length-address, zerobyte);
        printf("0x%X locations of ROM wrote.\n", address);
        return 0;
    }
    if (w_flag && a_flag)
    {
        unsigned int data = 0;
        address_t address = 0;
        if (filename) fp = fopen(filename, "r");
        printf("Writing ROM with contents of ascii file: %s from start 0x%X...\n", filename, start);
        while (fscanf(fp, "%2X", &data) == 1 && address<length)
        {
            writeROM(start+address, data);
            waitForWriteCycle(start+address, data);
            address++;
        }
        if (filename) fclose(fp);
        if (address<length)
            eraseROM(start+address, length-address, zerobyte);
        printf("0x%X locations of ROM wrote.\n", address);
        return 0;
    }
    if (v_flag && !a_flag)
    {
        address_t address = 0;
        data_t data = 0;
        data_t actual = 0;
        char result = 1;
        if (filename) fp = fopen(filename, "rb");
        if (!length) length = -1;
        printf("Verifying ROM with contents of binary file: %s from start 0x%X...\n", filename, start);
        while (!feof(fp) && address-length)
        {
            result = 0;
            fread(&data, sizeof(data_t), 1, fp);
            actual = readROM(start+address++);
            if (data != actual)
            {
                result = 1;
                break;
            }
        } 
        if (filename) fclose(fp);
        if (result)
            printf("verify error, found 0x%X at location 0x%X.\n", actual, --address);
        else
            printf("verify done.\n");
        return 0;
    }
    if (v_flag && a_flag)
    {
        address_t address = 0;
        unsigned int data = 0;
        unsigned int actual = 0;
        char result = 1;
        if (filename) fp = fopen(filename, "r");
        if (!length) length = -1;
        printf("Verifying ROM with contents of ascii file: %s from start 0x%X...\n", filename, start);
        while (fscanf(fp, "%2X", &data) == 1 && address-length)
        {
            result = 0;
            actual = readROM(start+address++);
            if (data != actual)
            {
                result = 1;
                break;
            }
        } 
        if (filename) fclose(fp);
        if (result)
            printf("verify error, found 0x%X at location 0x%X.\n", actual, --address);
        else
            printf("verify done.\n");
        return 0;
    }
    if (l_flag)
    {
        printf("ROM Software Data Protection %s.\n", (mode? "Enabled" : "Disabled"));
        setSDPMode(mode);
        return 0;
    }
    return 0;
}
